File: MusicManagerAPI/src/main/java/com/ldz/music/manager/service/MusicManagerService.java
package com.ldz.music.manager.service;

import com.ldz.music.manager.model.UserMusicStatus;
import com.ldz.converter.container.ConverterContainer;
import com.ldz.external.api.ExternalAPIClient;
import com.ldz.music.manager.constants.MusicTypes;
import com.ldz.music.manager.model.MusicType;
import com.ldz.music.manager.model.bo.MusicTypeBO;
import com.ldz.music.manager.model.bo.UserMusicStatusBO;
import com.ldz.music.manager.repository.MusicTypeRepository;
import com.ldz.music.manager.repository.UserMusicStatusRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Created by ldalzotto on 26/04/2017.
 */
@Transactional
@Service
public class MusicManagerService implements IMusicManagerService {

    @Autowired
    ExternalAPIClient externalAPIClient;

    @Autowired
    MusicTypeRepository musicTypeRepository;

    @Autowired
    UserMusicStatusRepository userMusicStatusRepository;

    @Autowired
    ConverterContainer converterContainer;

    @Override
    public void addUserMusicStatus(UserMusicStatusBO userMusicStatusBO) {
        UserMusicStatus userMusicStatus = converterContainer.convert(userMusicStatusBO, UserMusicStatus.class);

        if (userMusicStatus.getMusicTypes() != null) {
            userMusicStatus.getMusicTypes().iterator().forEachRemaining(musicType -> {
                musicType.setUserMusicStatus(userMusicStatus);
                musicTypeRepository.save(musicType);
            });
        }

        userMusicStatusRepository.save(userMusicStatus);
    }

    @Override
    public void deleteUserMusicStatusFromUsername(String username) {
        userMusicStatusRepository.deleteByUsername(username);
    }

    @Override
    public UserMusicStatusBO getMusicStatusFromUsername(String username) {
        UserMusicStatus userMusicStatus = userMusicStatusRepository.findByUsername(username);
        return converterContainer.convert(userMusicStatus, UserMusicStatusBO.class);
    }

    @Override
    public String getSourceurlFromUrlAndMusicType(String url, MusicTypes musicType) {
        ResponseEntity<String> sourceUrl;

        switch (musicType) {
            case SOUNDCLOUD:
                sourceUrl = externalAPIClient.getSoundcloudSourceurlFromRessource(url);
                return sourceUrl.getBody();
        }

        return null;
    }

    @Override
    public boolean postMusicFromUsernameAndSourceurl(String username, String sourceUrl) {
        //retrieve user
        UserMusicStatus userMusicStatus = userMusicStatusRepository.findByUsername(username);

        if (userMusicStatus == null) {
            return false;
        }

        //add url
        MusicTypeBO newMusic = new MusicTypeBO();
        newMusic.setMusicTypes(MusicTypes.SOUNDCLOUD);
        newMusic.setSourceUrl(sourceUrl);

        MusicType musicTypeToSave = converterContainer.convert(newMusic, MusicType.class);
        musicTypeToSave.setUserMusicStatus(userMusicStatus);

        musicTypeRepository.save(musicTypeToSave);

        return true;
    }

}

File: MusicManagerAPI/src/scala/com/ldz/music/manager/service/MusicManagerService.scala
package com.ldz.music.manager.service

import com.ldz.converter.container.ConverterContainer
import com.ldz.enumeration.ExternalMusicKey
import com.ldz.external.api.ExternalAPIClient
import com.ldz.music.manager.constants.MusicTypes
import com.ldz.music.manager.model.{MusicType, UserMusicStatus}
import com.ldz.music.manager.model.bo.{MusicTypeBO, UserMusicStatusBO}
import com.ldz.music.manager.repository.{MusicTypeRepository, UserMusicStatusRepository}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

import scala.collection.JavaConverters._

/**
  * Created by Loic on 20/06/2017.
  */
@Transactional
@Service
class MusicManagerService extends IMusicManagerService{

  @Autowired
  private val externalAPIClient: ExternalAPIClient = null

  @Autowired
  private val musicTypeRepository: MusicTypeRepository = null

  @Autowired
  private val userMusicStatusRepository: UserMusicStatusRepository = null

  @Autowired
  private val musicParametersFromRessourceService: IMusicParametersFromressource = null

  @Autowired
  private val converterContainer: ConverterContainer = null

  override def addUserMusicStatus(userMusicStatusBO: UserMusicStatusBO): Unit = {

    val userMusicStatus = converterContainer.convert(userMusicStatusBO, classOf[UserMusicStatus])



    userMusicStatus.getMusicTypes.asScala.foreach(musicType => {
                                                      musicType.setUserMusicStatus(userMusicStatus);
                                                      musicTypeRepository.save(musicType)
                                                      })
    userMusicStatusRepository.save(userMusicStatus)

  }

  override def deleteUserMusicStatusFromUsername(username: String): Unit = {
    userMusicStatusRepository.deleteByUsername(username)
  }

  override def getMusicStatusFromUsername(username: String): UserMusicStatusBO = {
    converterContainer.convert( userMusicStatusRepository.findByUsername(username), classOf[UserMusicStatusBO])
  }

  override def getMusicparametersFromUrlAndMusicType(url: String, musicType: MusicTypes.Value): MusicTypeBO = {
    musicParametersFromRessourceService.getMusicSourceFromRessource(musicType, url)
  }

  override def postMusicFromUsernameAndSourceurl(username: String, musicDetail: MusicTypeBO): MusicTypeBO = {

    val musicType =
        Option(userMusicStatusRepository.findByUsername(username)) match {
          case Some(userMusicStatus) => {
            val musicType = converterContainer.convert(musicDetail, classOf[MusicType])
            musicType.setType(musicType.getType)
            musicType.setUserMusicStatus(userMusicStatus)
            musicType.setMusicParameters(musicDetail.getMusicParameters.asJava)
            Option(musicTypeRepository.save(musicType))
          }
          case None => None
        }

    musicType match {
      case Some(mt) => converterContainer.convert(mt, classOf[MusicTypeBO])
      case None => throw new RuntimeException(s"An error occured wile saving ${musicDetail}")
    }

  }
}
